File: src/tools/glob.rs
//! `glob` tool: glob matching, returns matching paths (no contents).

use anyhow::{anyhow, Result};
use futures_util::future::FutureExt;
use ignore::WalkBuilder;
use serde_json::{json, Value};
use std::path::{Path, PathBuf};

use super::Tool;

const MAX_RESULTS: usize = 200;

pub fn tool() -> Tool {
    Tool {
        name: "glob",
        schema: json!({
            "type": "function",
            "function": {
                "name": "glob",
                "description": "Find files matching a glob pattern (e.g. **/*.rs). Honors .gitignore. Returns up to 200 paths.",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "pattern": { "type": "string" },
                        "path": { "type": "string", "description": "root dir; default cwd" }
                    },
                    "required": ["pattern"]
                }
            }
        }),
        exec: |args| async move { run(args).await }.boxed(),
    }
}

async fn run(args: Value) -> Result<String> {
    let pattern = args
        .get("pattern")
        .and_then(|v| v.as_str())
        .ok_or_else(|| anyhow!("glob: missing pattern"))?
        .to_string();
    let root = args
        .get("path")
        .and_then(|v| v.as_str())
        .map(|s| PathBuf::from(shellexpand::tilde(s).into_owned()))
        .unwrap_or_else(|| std::env::current_dir().unwrap_or_else(|_| PathBuf::from(".")));
    let g = globset::Glob::new(&pattern)
        .map_err(|e| anyhow!("invalid glob: {}", e))?
        .compile_matcher();

    let mut hits: Vec<String> = Vec::new();
    for entry in WalkBuilder::new(&root)
        .hidden(false)
        .git_ignore(true)
        .build()
    {
        let Ok(entry) = entry else { continue };
        if !entry.file_type().map(|ft| ft.is_file()).unwrap_or(false) {
            continue;
        }
        let p = entry.path();
        let rel = p.strip_prefix(&root).unwrap_or(p);
        let basename: &Path = p.file_name().map(Path::new).unwrap_or(p);
        if g.is_match(rel) || g.is_match(basename) {
            hits.push(p.display().to_string());
            crate::read_cache::mark_seen_by_search(p);
            if hits.len() >= MAX_RESULTS {
                break;
            }
        }
    }
    if hits.is_empty() {
        return Ok(format!(
            "(no matches for '{}' under {})\n",
            pattern,
            root.display()
        ));
    }
    Ok(hits.join("\n") + "\n")
}

File: src/tools/glob.ts
// `glob` tool: match files by pattern. Mirrors hip's tools/glob.rs.
// Uses node's built-in glob (Node 22+) which honors .gitignore via opts.

import { glob } from 'node:fs/promises';
import { resolve } from 'node:path';
import { type Tool, argString } from './types.js';

const MAX_RESULTS = 200;

export const globTool: Tool = {
  name: 'glob',
  spec: {
    type: 'function',
    function: {
      name: 'glob',
      description:
        "Find files matching a glob pattern (e.g. '**/*.ts', 'src/**/*.{js,jsx}'). Returns up to 200 paths sorted. Honors common ignores (node_modules, .git, dist, target).",
      parameters: {
        type: 'object',
        properties: {
          pattern: { type: 'string', description: 'glob pattern' },
          path: { type: 'string', description: 'root dir; default cwd' },
        },
        required: ['pattern'],
      },
    },
  },
  async run(args) {
    const pattern = argString(args, 'pattern');
    if (!pattern) throw new Error('glob: missing pattern');
    const root = argString(args, 'path') ?? process.cwd();
    const cwd = resolve(root);
    const exclude = (p: string) =>
      /\/(node_modules|\.git|target|dist|build|\.next|\.venv|\.cache)(\/|$)/.test(p);
    const matches: string[] = [];
    for await (const p of glob(pattern, { cwd })) {
      if (exclude(p)) continue;
      matches.push(p);
      if (matches.length >= MAX_RESULTS) break;
    }
    matches.sort();
    if (matches.length === 0) return `[no matches for ${pattern} under ${cwd}]\n`;
    const truncated = matches.length === MAX_RESULTS ? ` (truncated at ${MAX_RESULTS})` : '';
    return `${matches.length} match(es)${truncated}:\n${matches.join('\n')}\n`;
  },
};
